ocaml/xapi/xapi_vtpm.ml

```diff
@@ -12,6 +12,30 @@
    GNU Lesser General Public License for more details.
  *)
 
+let bad_power_state ~vm ~expected ~actual =
+  let p2s = Record_util.power_to_string in
+  let vm2s = Ref.string_of in
+  raise
+    Api_errors.(
+      Server_error (vm_bad_power_state, [vm2s vm; p2s expected; p2s actual])
+    )
+
+let assert_operation_valid ~__context ~self ~op =
+  let vm = Db.VTPM.get_VM ~__context ~self in
+  (* currently every VTPM has an associated VM *)
+  assert (Db.is_valid_ref __context vm) ;
+  let state = Db.VM.get_power_state ~__context ~self:vm in
+  let ok = () in
+  match op with
+  | `destroy when state = `Halted ->
+      ok
+  | `set_contents when state = `Halted ->
+      ok
+  | `get_contents when state = `Halted ->
```

Comment:
I think API users are not meant to use set/get contents at all, shouldn't it be purely a XAPI operation?
(and currently swtpm will only use it when the VM is running).
In general API users should use export/import VM metadata instead.

Maybe we should mark set/get contents as internal only so you cannot call it with *any* API role unless you are local root and XAPI itself (or swtpm).

Reply:
The allowed operations have no direct connection with API calls - they are connected by convention and form a set of names. So we can put anything here we like but I mirrored the API calls. We can still make the get/set contents API calls private - which would leave only create and destroy as operations.